### DISCUSSION TITLE
More than one edge data variable

### DISCUSSION THREAD

**Discussion by dschult** [networkx/networkx#7207] [General]
[I am posting this text from a recent email so it can get a wider audience and others with similar question can find it here. Thanks Peter de KAM for the question and feel free to add to this discussion here.]
Hello,

I am working with fuzzy cognitive modeling of actors working in inter-organizational collaboration (think strategic alliances and partnerships).  I have a relatively short model represented as a fuzzy digraph with 42 nodes.  I have been able to load the digraph into NetworkX and have achieved interesting results by looking at various aspects of centrality.

However, I recently had a revelation that I could potentially account for more than one aspect of the relationships between nodes using tuples.  For instance, I can (conventionally) account for the fuzzy influence one node has on another node.  It also appears as though I can account for the amount of time it takes to achieve this influence as well as the fuzzification of the characteristics of the nodes. Several other aspects of cognition could potentially be used to characterize the relationships between nodes.  Is this a case for the use of multidigraph where there are multiple arcs between nodes?

The problem is that I am not at all confident that this is a "valid" application of NetworkX to what I think I want to accomplish.  I can intuitively build a picture of the effect of one vector on a node.  I can also mentally picture sets of vectors interacting with each other on a node as represented in a "tuple" I am currently using the following lines of code to enter the digraph for analysis:

```python
import networkx as nx
import numpy as np
import matplotlib.pyplot

# Create directed graph
G = nx.DiGraph()

# Add nodes
G.add_nodes_from(["A", "B", "C", "Y"])

# Add weighted edges
G.add_weighted_edges_from([("A", "B", 0.52)])
G.add_weighted_edges_from([("A", "C", 0.5)])
G.add_weighted_edges_from([("A", "H", 1)])
```
 
I've attached a copy of my latest code for your reference.
 
-  How do I use a tuple to describe multiple aspects of a vector?  Or, is this a case for using Multidigraphs?
-  How do I represent these "complex" vectors in Matplotlib?
-  I understand how I can create a Monte Carlo simulation using one edge weight...  How do I create a Monte Carlo simulation with multiple (complex) edge weights?  Or, do I have to create a new single edge weight from the interaction between arcs/vectors before running the simulation? 
-  How do multidigraphs account for any interaction between arcs from one vertex to the next?
-  Would I have to create subgraphs for the multiple arcs in a multidigraph to account for any interactions between arcs?
-  If I compute various degrees of centrality with the new complex arcs/vectors, what am I actually measuring?
-  If I use Multidigraphs to represent multiple arcs from one vertex to another do my concerns go away?
-  The Reference/Algorithms page in the NetworkX documentation lists quite a number of different ways to analyze graphs, given what I have shared with you so far, what types of analysis should I think of using (ie Chordal graphs, efficiency, CCPA scores, etc.??
-  if I want to get valid results, are there any other questions I need to be asking?

I hope I have adequately described my thinking process and you can shed a little light on how I can achieve my goal. of creating a fuzzy social simulation of the cognitive architecture of inter-organizational collaboration.  I finished my doctoral program 18 months ago when I was 66 after very full careers in the military and in the corporate world so, “some” of this has changed since my first class in computer science using punch cards and FORTRAN in 1977!! So, I would greatly appreciate any and all support you may be able to provide as I write my job market paper. Please feel free to refer me to others who may have an active interest in my research goals.

Some context:
The "problem" of inter-organizational collaboration is that most of them fail within two years costing the economy trillions in lost opportunity.  However, 99% of the research in the literature has been from the standpoint of transaction cost economics which has not answered the question of why inter-organizational collaboration isn't not working. So, I'm trying to look at it from the standpoint of cognition and, hopefully, come up with a method that recognizes the dynamic understructure of cognition through the use of network science...  It's my belief there could be several papers and I am more than willing to team write papers with people who are as motivated to think out of the box as I am.

I've attached a copy of the Java code I have written so far to this note for others to use as a frame of reference.
Cheers

//Peter

**Comment by dschult** [comment]
Typically, we have more than one data variable for edges. The algorithms often use only one variable at a time. So we allow users to specify which edge attribute to use when they call the function for their algorithm.  For example,
```python
spl_trade = nx.shortest_path_length(G, u, v, weight="trade")
spl_travel = nx.shortest_path_length(G, u, v, weight="air_traffic")
```
The edge data attributes can be moved between NetworkX and other data frameworks with functions like: [nx.set_edge_attribute](https://networkx.org/documentation/latest/reference/generated/networkx.classes.function.set_edge_attributes.html)
The edge data attributes are stored on each edge as a dict keyed by attribute name to the value of that attribute. So to set an attribute for a specific edge between `u` and `v`, use `G[u][v]["correlation"]=0.24`.  
To move an entire pandas DataFrame with two columns for the source and target node  respectively and other columns of data, use [nx.from_pandas_edgelist](https://networkx.org/documentation/latest/reference/generated/networkx.convert_matrix.from_pandas_adjacency.html).  Similarly from a Numpy array storing an adjacency matrix of attribute values, use [nx.from_numpy_array](https://networkx.org/documentation/latest/reference/generated/networkx.convert_matrix.from_numpy_array.html).

Node attributes and information are stored similarly: as a dict keyed by attribute name to the attribute value. So, `G.nodes[node]["size"]=5` is possible, but its probably better to use the interface for numpy or pandas data storage when moving the data for large networks. And there are `nx.set_node_attributes` and `nx.get_node_attributes` too. 

More [info on manipulating data](https://networkx.org/documentation/latest/reference/convert.html) is available in the docs too.

You can use more than one edge variable to compute e.g. the shortest path in each variable and manipulate them separately or combine them after they are created.  Or, you can build funky data values that are e.g. tuples of variable values. The shortest path algorithm uses `<` and `>` to find the shortest path, and tuples sort based on first element with second element only if the first elements are equal, and so on.  So, if you make your attribute values into tuples of values, you can find the shortest path using those tuples of values.

Betweenness_centrality can be used just like shortest_path function because it is just computing shortest paths. But some other centrality measures like eigenvector_centrality require numerical values for the edge weights so that tuples will not work (multiplication is not defined between tuples for example).  The system is intended to be very flexible. Hopefully it is also intuitive enough to allow you to play some with your data.

Cheers,
Dan

**Reply by peterdekam** [reply]
It's been a month since I last looked at this, and I have lost the "bubble."  How can I look at the interaction between multiple edge elements? For instance, I would like to have an edge weight "strength of relationship" and another edge weight "time."  Your advice seems to indicate that I need to account for one edge weight at a time. Do I need to create a third variable that represents the interaction between the "strength of relationship" and "time"?  Is it better to store the adjacency matrix, edge weight, and time as elements of an array versus using a tuple?  I don't see the path from my initial code represented below to your description above...

# Create directed graph
G = nx.DiGraph()

# Add nodes
G.add_nodes_from(["A", "B", "C", "Y"])

# Add weighted edges
G.add_weighted_edges_from([("A", "B", 0.52)])
G.add_weighted_edges_from([("A", "C", 0.5)])
G.add_weighted_edges_from([("A", "H", 1)])

**Reply by rossbar** [reply]
Copying and pasting the chunk that I think is most relevant to your question for visibility:

> You can use more than one edge variable to compute e.g. the shortest path in each variable and manipulate them separately or combine them after they are created. Or, you can build funky data values that are e.g. tuples of variable values. The shortest path algorithm uses < and > to find the shortest path, and tuples sort based on first element with second element only if the first elements are equal, and so on. So, if you make your attribute values into tuples of values, you can find the shortest path using those tuples of values.

**Reply by peterdekam** [reply]
Thanks very much for the reply...  The problem I have is related to understanding the relationship between the way in which I created the digraph... I grew my digraph with individual nodes and edges using the following format:
`G.add_weighted_edges_from([("A", "B", 0.52)])`
As I understand it, tuples are formed by using parentheses.  As a result, I think I can use the following convention:
`G.add_weighted_edges_from([("A", "B", (0.52, 64))])`
Do I understand this correctly?  If I didn't, how should it be represented?  
Do I also understand that degree centrality and other measures of centrality calculate one tuple element at a time...  Is this correct? When I print the results, is the output rendered one element at a time?
I don't mean to be dense, but I started learning Python four months ago by reading the documentation and other education-oriented websites, so I have limited prior knowledge and need things explained rather simplistically. 
Cheers

**Reply by dschult** [reply]
You can use the `weight` argument to give a variable name to the weight value that you add.
```python
G = nx.Graph()
G.add_weighted_edges_from([("A", "H", 1)], weight="time")
G.add_weighted_edges_from([("A", "H", 103.5)], weight="strength")
```

Then for shortest_path_length, you can provide a `weight` keyword arg as a function that computes the weight based on variables held in the data dictionary.

```python
def wt(u, v, dd):
    return dd['time'] - dd['strength']

nx.shortest_path_length(G, 'A', 'H', weight=wt)  # output:  -102.5
dict(nx.shortest_path_length(G, weight=wt))  # output: {'A': {'A': 0, 'H': -102.5}, 'H': {'H': 0, 'A': -102.5}}
```
For MultiGraph the weight function should have inputs `wt(u, v, edge_key, dd)`.

**Reply by dschult** [reply]
OK... I see your post now.. :}
Yes, 2-tuples like (3, 53) and (2, 64) are compared using the first element and only if it is the same do the 2nd elements get compared. So:
`(3, 53) > (2, 64) is True` and `(2, 53) > (2, 64) is False`.
